pkg/dockerregistry/server/auth.go

```diff
@@ -105,69 +100,64 @@ func (app *App) Auth(options map[string]interface{}) (registryauth.AccessControl
 }
 
 // Error returns the internal error string for this authChallenge.
-func (ac *authChallenge) Error() string {
+func (ac *tokenAuthChallenge) Error() string {
 	return ac.err.Error()
 }
 
-// SetHeaders sets the basic challenge header on the response.
-func (ac *authChallenge) SetHeaders(req *http.Request, w http.ResponseWriter) {
+// SetHeaders sets the basic and bearer challenge headers on the response.
+func (ac *tokenAuthChallenge) SetHeaders(req *http.Request, w http.ResponseWriter) {
 	// WWW-Authenticate response challenge header.
-	// See https://tools.ietf.org/html/rfc6750#section-3
-	str := fmt.Sprintf("Basic realm=%s", ac.realm)
+	// See https://docs.docker.com/registry/spec/auth/token/#/how-to-authenticate and https://tools.ietf.org/html/rfc6750#section-3
+	str := fmt.Sprintf("Basic realm=%s", ac.basicRealm)
 	if ac.err != nil {
-		str = fmt.Sprintf("%s,error=%q", str, ac.Error())
+		str += fmt.Sprintf(",error=%q", ac.Error())
 	}
 	w.Header().Set("WWW-Authenticate", str)
-}
 
-// Error returns the internal error string for this authChallenge.
-func (ac *tokenAuthChallenge) Error() string {
-	return ac.err.Error()
-}
-
-// SetHeaders sets the bearer challenge header on the response.
-func (ac *tokenAuthChallenge) SetHeaders(req *http.Request, w http.ResponseWriter) {
-	// WWW-Authenticate response challenge header.
-	// See https://docs.docker.com/registry/spec/auth/token/#/how-to-authenticate and https://tools.ietf.org/html/rfc6750#section-3
-	str := fmt.Sprintf("Bearer realm=%q", ac.realm)
+	str = fmt.Sprintf("Bearer realm=%q", ac.bearerRealm)
+	if ac.err != nil {
+		str += fmt.Sprintf(",error=%q", ac.Error())
+	}
 	if ac.service != "" {
 		str += fmt.Sprintf(",service=%q", ac.service)
 	}
-	w.Header().Set("WWW-Authenticate", str)
+	w.Header().Add("WWW-Authenticate", str)
 }
 
 // wrapErr wraps errors related to authorization in an authChallenge error that will present a WWW-Authenticate challenge response
 func (ac *AccessController) wrapErr(ctx context.Context, err error) error {
+	// Auto-detect scheme/host from request
+	req, reqErr := dcontext.GetRequest(ctx)
+	if reqErr != nil {
+		return reqErr
+	}
+	scheme, host := httprequest.SchemeHost(req)
+	tokenRealmCopy := *ac.tokenRealm
```

Comment:
Either it cannot be nil (and this case you can remove checks like `if ac.tokenRealm == nil {`) or you should handle nil.